crates/red_knot_test/src/lib.rs

```diff
@@ -65,31 +72,40 @@ fn run_test(test: &parser::MarkdownTest) -> Result<(), Failures> {
         );
         let full_path = workspace_root.join(file.path);
         db.write_file(&full_path, file.code).unwrap();
-        system_paths.push(full_path);
+        paths.push(AbsoluteLineNumberPath {
+            path: full_path,
+            starting_line_number: file.starting_line_number,
+        });
     }
 
-    let mut failures = BTreeMap::default();
+    let mut failures = Vec::with_capacity(paths.len());
+    paths.sort_by(|a, b| a.starting_line_number.cmp(&b.starting_line_number));
 
-    for path in system_paths {
-        let file = system_path_to_file(&db, path.clone()).unwrap();
+    for contextual_path in paths {
+        let file = system_path_to_file(&db, contextual_path.path.clone()).unwrap();
         let parsed = parsed_module(&db, file);
 
         // TODO allow testing against code with syntax errors
         assert!(
             parsed.errors().is_empty(),
             "Python syntax errors in {}, {:?}: {:?}",
             test.name(),
-            path,
+            contextual_path.path,
             parsed.errors()
         );
 
-        matcher::match_file(&db, file, check_types(&db, file)).unwrap_or_else(|line_failures| {
-            failures.insert(path, line_failures);
-        });
+        match matcher::match_file(&db, file, check_types(&db, file)) {
+            Ok(()) => {}
+            Err(line_failures) => {
+                failures.push(line_failures.offset_errors(contextual_path.starting_line_number));
+            }
+        }
     }
+
     if failures.is_empty() {
         Ok(())
     } else {
+        failures.shrink_to_fit();
```

Comment:
I don't think calling `shrink_to_fit` gives us much here. `shrink_to_fit` is mainly useful if you have a long-living `Vec` to avoid wasting memory. The failures are consumed right away (and if not, let the caller handle it) and then dropped.